### PR TITLE
add subslice support in drop ladder for array

### LINKED ISSUE #34708: [MIR] double drop with slice patterns
## Version

```
$ rustc -V
rustc 1.11.0-dev (42b7c32ac 2016-07-02)
```
## STR

``` Rust
#![feature(slice_patterns, rustc_attrs)]

struct NoisyDrop;
impl Drop for NoisyDrop {
    fn drop(&mut self) { println!("splat!"); }
}

#[rustc_mir]
fn main() {
    let [x] = [NoisyDrop];
}
```
## Expected Result

`NoisyDrop` gets dropped once
## Actual Result

`NoisyDrop` is dropped twice

### PR DESCRIPTION
**Pull request by mikhail-m1**
second step for fix #34708

### CODE REVIEW

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -1279,6 +1278,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     );
                 }
             }
+            Write(WriteKind::Move) |
```
Comment (arielb1) [inline src/librustc_mir/borrow_check/mod.rs]: Why this change?

Comment (arielb1) [reply src/librustc_mir/borrow_check/mod.rs]: Why did you make this change.

File: src/librustc_mir/util/elaborate_drops.rs [hunk 2]
```diff
@@ -88,14 +88,19 @@ pub trait DropElaborator<'a, 'tcx: 'a> : fmt::Debug {
     fn param_env(&self) -> ty::ParamEnv<'tcx>;
 
     fn drop_style(&self, path: Self::Path, mode: DropFlagMode) -> DropStyle;
-    fn get_drop_flag(&mut self, path: Self::Path) -> Option<Operand<'tcx>>;
-    fn clear_drop_flag(&mut self, location: Location, path: Self::Path, mode: DropFlagMode);
+    fn get_drop_flags(&mut self, path: Self::Path) -> Option<Operand<'tcx>>;
+    fn clear_drop_flag(&mut self,
+                       location: Location, path:
```
Comment (arielb1) [inline src/librustc_mir/util/elaborate_drops.rs]: nit: path should go in the next line

File: src/librustc_mir/transform/elaborate_drops.rs [hunk 3]
```diff
@@ -257,18 +268,37 @@ impl<'a, 'b, 'tcx> DropElaborator<'a, 'tcx> for Elaborator<'a, 'b, 'tcx> {
...
-                &Projection {
-                    elem: ProjectionElem::ConstantIndex{offset, min_length: _, from_end: true}, ..
-                } => size - offset == index,
-                _ => false
-            }
-        })
+    fn array_subpaths(&self, path: Self::Path, size: u64)
+                      -> Vec<(Place<'tcx>, Option<Self::Path>, Option<Local>)> {
+        if dataflow::move_path_children_matching(self.ctxt.move_data(), path, |_| {
+                assert!(size <= (u32::MAX as u64),
```
Comment (arielb1) [inline src/librustc_mir/transform/elaborate_drops.rs]: can you move the assertion out of the loop?

File: src/librustc_mir/transform/elaborate_drops.rs [hunk 4]
```diff
@@ -257,18 +268,37 @@ impl<'a, 'b, 'tcx> DropElaborator<'a, 'tcx> for Elaborator<'a, 'b, 'tcx> {
...
+                        "move out check doesn't implemented for array bigger then u32");
+                true
+            }).is_none() {
+            return vec![];
+        }
+
+        let size = size as u32;
+        let flags =  self.ctxt.array_items_drop_flags.get(&path);
```
Comment (arielb1) [inline src/librustc_mir/transform/elaborate_drops.rs]: nit: stray double space
```
flags =  self.ctxt
       ^^
```

**Comment by arielb1** [comment]
Do you handle the case where either individual indexes or subslices are moved out, aka:
```Rust
fn subslice_pattern_test(a: &Allocator, arg: bool) {
    let a = [a.alloc(), a.alloc(), a.alloc(), a.alloc(), a.alloc()];
    if arg {
        let[.., _x, _] = a;
    } else {
        let[_, _y..] = a;
    }
}
```

File: src/librustc_mir/util/elaborate_drops.rs [hunk 5]
```diff
@@ -88,14 +88,19 @@ pub trait DropElaborator<'a, 'tcx: 'a> : fmt::Debug {
     fn param_env(&self) -> ty::ParamEnv<'tcx>;
 
     fn drop_style(&self, path: Self::Path, mode: DropFlagMode) -> DropStyle;
-    fn get_drop_flag(&mut self, path: Self::Path) -> Option<Operand<'tcx>>;
-    fn clear_drop_flag(&mut self, location: Location, path: Self::Path, mode: DropFlagMode);
+    fn get_drop_flags(&mut self, path: Self::Path) -> Option<Operand<'tcx>>;
```
Comment (arielb1) [inline src/librustc_mir/util/elaborate_drops.rs]: this still can't return multiple drop flags, so why did you rename it?

File: src/librustc_mir/util/elaborate_drops.rs (hunk 2 above)
Comment (arielb1) [reply src/librustc_mir/util/elaborate_drops.rs]:
```
location: Location,
path: Self::Path
```

**Comment by arielb1** [comment]
Another test:
```Rust
fn subslice_pattern_test(a: &Allocator, arg: bool, arg2: bool) {
    let a = [a.alloc(), a.alloc(), a.alloc(), a.alloc(), a.alloc()];
    if arg2 {
        drop(a);
        return;
    }

    if arg {
        let[.., _x, _] = a;
    } else {
        let[_, _y..] = a;
    }
}
```

File: src/librustc_mir/transform/elaborate_drops.rs [hunk 6]
```diff
@@ -389,6 +492,7 @@ impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
                 }
             };
 
+
```
Comment (arielb1) [inline src/librustc_mir/transform/elaborate_drops.rs]: nit: stray newline

File: src/librustc_mir/transform/elaborate_drops.rs [hunk 7]
```diff
@@ -257,18 +268,38 @@ impl<'a, 'b, 'tcx> DropElaborator<'a, 'tcx> for Elaborator<'a, 'b, 'tcx> {
...
+                    _ => false
+                }
+            );
+            let place = &self.ctxt.move_data().move_paths[path].place;
+            (place.clone().elem(ProjectionElem::ConstantIndex{
+                offset: i,
+                min_length: size,
+                from_end: false
+            }),
+            move_path,
```
Comment (arielb1) [inline src/librustc_mir/transform/elaborate_drops.rs]: So reading this, I'm afraid that this can cause a drop to be run with a "wrong" `MovePath` - a `Subslice` when the drop can also be accessed through a `ConstantIndex` (or vice versa), which might cause the "unconditional" drop code to trigger when it shouldn't, as in

```Rust
let mut x = [a.alloc(), a.alloc(), a.alloc()];
// The MP for x[1..3] is now initialized, x[1] is now uninitialized
let [_, _, _x] = x;
// this shouldn't rely on any drop flags, and just drop x[0] & x[1]
x = [a.alloc(), a.alloc(), a.alloc()];
let [_, .._x] = x;
// The MP for x[1..3] is now uninitialized, x[1] is now initialized
// end of scope drop for x - should drop x[0] without any drop flags
```

Comment (arielb1) [reply src/librustc_mir/transform/elaborate_drops.rs]: And in some cases you might need the drop flags:
```Rust
let alloc = || (a.alloc(). a.alloc());
let mut x = [alloc(), alloc(), alloc()];
// The MP for x[1..3] is now initialized, x[2].0 is now uninitialized
let [_, _, (_x,_)] = x;
// this shouldn't rely on any drop flags, and just drop x[0], x[1], and x[2].1
x = [a.alloc(), a.alloc(), a.alloc()];
let [_, .._x] = x;
// The MP for x[1..3] is now uninitialized, x[1].1 is now initialized
// end of scope drop for x - should drop x[0] without any drop flags
```

So I think you *do* need to generate move paths for all the array elements when you see a subslice, to have a place to focus (similarly, you want to normalize the move paths so that an access to `array[-2 of 3]` is the same as an access to `array[1 of 3]`)

Comment (arielb1) [reply src/librustc_mir/transform/elaborate_drops.rs]: At least, this is how it appears to me. There might be a better solution but it elides me.

Comment (arielb1) [reply src/librustc_mir/transform/elaborate_drops.rs]: I'll go back to european time and no meetings in Monday (I'm currently in the US for the Mozilla All Hands), so if you need any help I'll go on IRC then.

File: src/librustc_mir/transform/elaborate_drops.rs [hunk 8]
```diff
@@ -257,18 +268,38 @@ impl<'a, 'b, 'tcx> DropElaborator<'a, 'tcx> for Elaborator<'a, 'b, 'tcx> {
...
-                } => offset == index,
-                &Projection {
-                    elem: ProjectionElem::ConstantIndex{offset, min_length: _, from_end: true}, ..
-                } => size - offset == index,
-                _ => false
-            }
-        })
+    fn array_subpaths(&self, path: Self::Path, size: u64)
+                      -> Vec<(Place<'tcx>, Option<Self::Path>, Option<Local>)> {
+        if dataflow::move_path_children_matching(self.ctxt.move_data(), path,
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/elaborate_drops.rs]: Nit: speaking as someone who doesn't know this code that well, I'd appreciate some comments here. =)

I guess that this is just a "trivial case" check, however? i.e., if we never touched any of the indices in this array, then we there is nothing to do here?

**Comment by arielb1** [comment]
I don't think I can finish handling this PR.

r? @nikomatsakis

**Comment by arielb1** [comment]
@nikomatsakis could you try to get this working with @mikhail-m1 ?

**Comment by nikomatsakis** [comment]
@arielb1 @mikhail-m1 yep!

**Comment by kennytm** [comment]
Hi @mikhail-m1, just to check if you're still working on this. Looks like @arielb1 has [some questions](https://github.com/rust-lang/rust/pull/46686#pullrequestreview-83681429) for you too!

**Comment by mikhail-m1** [comment]
@kennytm Hi, @arielb1 showed me serious issue in my current solution, I will be back to finding new approach next week.

**Comment by arielb1** [comment]
@mikhail-m1

Cool! I'll be sure to review your PR when you get it done.

**Comment by shepmaster** [comment]
Ping from triage @mikhail-m1! It's been over a week since we last heard from you — will you still be able to make the needed revisions?

**Comment by mikhail-m1** [comment]
@shepmaster sorry, I am still working on it but don't have enough time to finish, hope I will do it in next few days
